rfcs/plugins.md

```diff
@@ -0,0 +1,349 @@
+following properties:
+  - `optionsSchema`: A Joi schema to validate the request type.
+  - `clientValidation`: An object with the keys:
+    - `minArgs`: minimum number of arguments accepted
+    - `maxArgs`: maximum number of arguments accepted
+    - `legalToChainFrom`: a list of terms that can be chained before
+      this request method. The special value `"/"` indicates the
+      method is available directly on the `horizon`
+      instance. `"collection"` indicates it's chainable from a
+      collection.
+    - `nullable`: a boolean determining whether any arguments can be
+      null
+  - `handler`: A function that receives a validated request, the
+    current plugin state
+
+The schema provided just validates the contents of the `options`
+field. So for a schema like:
+
+```js
+requests: {
+  myRequestType: {
+    optionsSchema: {
+      collection: Joi.string(),
+      findAll: Joi.array(Joi.any()),
+      myRequestType: Joi.array().ordered(
+          Joi.number(), Joi.boolean().default(false))
+    },
+    clientValidation: {
+      minArgs: 1,
+      maxArgs: 2,
+      nullable: true,
+      legalToChainFrom: [ '/', 'collection', 'findAll' ],
+    },
+    handler: (requestOptions, pluginState) => { ... },
+  }
+}
+```
+
+This will inform the client that it needs to add a method that can be
+called like:
+
+``` js
+horizon.myRequestType(33, true)
+horizon('abc').myRequestType(12)
+horizon('abc').myRequestType("string") // server will error
+horizon('abc').findAll({ thing: 123 }).myRequestType(22)
+// The following will throw an exception when they are called:
+horizon.myRequestType() // not enough arguments
+horizon.myRequestType(1,2,3) // too many arguments
+horizon('abc').find({ id: 12 }).myRequestType(55) // not chainable from `find`
+```
+
+A raw request might look like:
+
+```js
+{
+    "request_id": 126,
+    "type": "myRequestType",
+    "options": {
+        "collection": "abc",
+        "myRequestType": [ 12 ]
+    }
+}
+```
+
+** Note **
+Although the request is chained off built in methods in Horizon, it is entirely up to the plugin to do something with the `options.collection`
+key here, as well as any other keys that it allows itself to be chained from.
+
+The plugin is responsible for responding to `end_subscription` requests, and for sending `state: complete` to the client if no more results are available for the current request. See the [protocol document](https://github.com/rethinkdb/horizon/blob/next/docs/protocol.md) for details.
+
+### Activate
+
+The `activate` plugin property must be present and contain a function
+that accepts an object with the following properties:
+  - `config`: config options from the config.toml section for the
+    plugin. This includes user overrides which are not specified in
+    the `config` plugin property.
+  - `env`: an object with relevant environment variables (defined in
+    [environment](#environment))
+  - `rdbConn`: a function that returns an open rethinkdb connection
```

Comment:
This is in general a hard thing, since we don't provide resumable changefeeds. I'd say we should expect dispose() to be called before activate is called a second time